File: src/devmemory/watch/state.py
"""Watermark persistence for the watch daemon.

Tracks, per conversation, how many messages have already been saved and which
DevMemory session they were saved to — so a restart never re-saves old turns
and every turn of one conversation lands in the same session.

Stored as JSON at ``~/.devmemory/watch_state.json``. Corrupt/missing state is
treated as empty (safe: at worst we re-save recent turns once).
"""

from __future__ import annotations

import json
from pathlib import Path


class WatchState:
    def __init__(self, path: Path | None = None) -> None:
        self.path = path or (Path.home() / ".devmemory" / "watch_state.json")
        self._data: dict = {}
        self._load()

    def _load(self) -> None:
        try:
            self._data = json.loads(self.path.read_text(encoding="utf-8"))
        except (OSError, ValueError):
            self._data = {}
        if not isinstance(self._data, dict):
            self._data = {}
        self._data.setdefault("conversations", {})

    def save(self) -> None:
        self.path.parent.mkdir(parents=True, exist_ok=True)
        tmp = self.path.with_suffix(".json.tmp")
        tmp.write_text(json.dumps(self._data, indent=2) + "\n", encoding="utf-8")
        tmp.replace(self.path)

    # ── per-conversation watermark ──────────────────────────────────────────

    def saved_count(self, key: str) -> int:
        """How many messages of this conversation have already been saved."""
        return int(self._data["conversations"].get(key, {}).get("saved_count", 0))

    def session_id(self, key: str) -> str | None:
        return self._data["conversations"].get(key, {}).get("session_id")

    def record(self, key: str, saved_count: int, session_id: str | None) -> None:
        entry = self._data["conversations"].setdefault(key, {})
        entry["saved_count"] = saved_count
        if session_id:
            entry["session_id"] = session_id

File: clients/npm/src/watch/state.js
// Watermark persistence — mirrors the Python daemon's WatchState.
//
// Tracks, per conversation, how many messages have already been saved and which
// DevMemory session they landed in, so a re-scan never re-saves old turns and
// every turn of one conversation stays in the same session. Stored at
// ~/.devmemory/watch_state.json in the SAME format the Python daemon uses, so
// the two clients share one watermark and never double-save each other's work.

import { mkdirSync, readFileSync, renameSync, writeFileSync } from "node:fs";
import { homedir } from "node:os";
import { dirname, join } from "node:path";

const STATE_PATH = join(homedir(), ".devmemory", "watch_state.json");

export class WatchState {
  constructor(path = STATE_PATH) {
    this.path = path;
    this._data = { conversations: {} };
    try {
      const parsed = JSON.parse(readFileSync(this.path, "utf8"));
      if (parsed && typeof parsed === "object") {
        this._data = parsed;
        if (!this._data.conversations || typeof this._data.conversations !== "object") {
          this._data.conversations = {};
        }
      }
    } catch {
      /* missing / corrupt → empty (safe: at worst re-save recent turns once) */
    }
  }

  savedCount(key) {
    const entry = this._data.conversations[key];
    return entry && Number.isFinite(entry.saved_count) ? Number(entry.saved_count) : 0;
  }

  sessionId(key) {
    const entry = this._data.conversations[key];
    return (entry && entry.session_id) || null;
  }

  record(key, savedCount, sessionId) {
    const entry = this._data.conversations[key] || {};
    entry.saved_count = savedCount;
    if (sessionId) entry.session_id = sessionId;
    this._data.conversations[key] = entry;
  }

  save() {
    mkdirSync(dirname(this.path), { recursive: true });
    const tmp = `${this.path}.tmp`;
    writeFileSync(tmp, `${JSON.stringify(this._data, null, 2)}\n`, "utf8");
    renameSync(tmp, this.path);
  }
}
